syntax = "proto3";

enum Action {
	DEFAULT = 0;
	UNAUTHORIZED = 1;
	MAINTENANCE = 2;
	GEOIP_BLOCKING = 3;
}

message UpdatedTitle {
	Title title = 1;
	uint32 chapter_id = 2;
	string chapter_name = 3;
	string chapter_sub_title = 4;
	bool is_latest = 5;
}

message UpdatedTitleGroup {
	repeated UpdatedTitle titles = 2;
}

message WebHomeView {
	repeated UpdatedTitleGroup groups = 2;
}

message Popup {
	string subject = 1;
	string body = 2;
}

message Title {
	enum Language {
		ENGLISH = 0;
		SPANISH = 1;
	}
	uint32 title_id = 1;
	string name = 2;
	string author = 3;
	string portrait_image_url = 4;
	string landscape_image_url = 5;
	uint32 view_count = 6;
	Language language = 7;
}

message Chapter {
	uint32 title_id = 1;
	uint32 chapter_id = 2;
	string name = 3;
	string sub_title = 4;
	string thumbnail_url = 5;
	uint32 start_timestamp = 6;
	uint32 end_timestamp = 7;
}

message TitleDetailView {
	enum UpdateTiming {
		NOT_REGULARLY = 0;
		MONDAY = 1;
		TUESDAY = 2;
		WEDNESDAY = 3;
		THURSDAY = 4;
		FRIDAY = 5;
		SATURDAY = 6;
		SUNDAY = 7;
		DAY = 8;
	}
	Title title = 1;
	string title_image_url = 2;
	string overview = 3;
	string background_image_url = 4;
	uint32 next_timestamp = 5;
	UpdateTiming update_timing = 6;
	string viewing_period_description = 7;
	string non_appearance_info = 8;
	repeated Title recommended_titles = 12;
	bool is_simul_release = 14;
	// 27 is title ids for other languages
	ChaptersView chapters = 28;
}

message ChaptersView {
	repeated Chapter first_chapter_list = 2;
	repeated Chapter unavailable_chapter_list = 3;
	repeated Chapter last_chapter_list = 4;
}

// Different language versions of a title
message TitleVariants {
	string name = 1;
	repeated Title title = 2;
}

message AllTitlesView {
	repeated TitleVariants title_variants = 1;
}

message SuccessResult {
	oneof result {
		TitleDetailView title_detail = 8;
		WebHomeView web_home_view = 11;
		AllTitlesView all_titles = 25;
	}
}

message ErrorResult {
	Action action = 1;
	Popup english_popup = 2;
	Popup spanish_popup = 3;
	string debug_info = 4;
}

message Response {
	SuccessResult success_result = 1;
	ErrorResult error_result = 2;
}
